sdk/lib/cmlib/hiveinit.c

```diff
@@ -357,98 +476,651 @@ HvpInitializeFlatHive(
+        LogHeader->Cluster = 1;
+        LogHeader->Length = HiveSize;
+        LogHeader->CheckSum = HvpHiveHeaderChecksum(LogHeader);
+
+        /*
+         * Acknowledge that we have resuscitated
+         * the header.
+         */
+        HeaderResuscitated = TRUE;
+        DPRINT1("HvpRecoverHeaderFromLog(): Header has been resuscitated, triggering self-heal mode!\n");
+    }
+
+    /*
+     * Tag this log header as a primary hive before
+     * writing it to the hive.
+     */
+    LogHeader->Type = HFILE_TYPE_PRIMARY;
+
+    /*
+     * If we have not made attempts of recovering
+     * the header due to log corruption then we
+     * have to compute the checksum. This is
+     * already done when the header has been resuscitated
+     * so don't try to do it twice.
+     */
+    if (!HeaderResuscitated)
+    {
+        LogHeader->CheckSum = HvpHiveHeaderChecksum(LogHeader);
+    }
+
+    /* Write the header back to hive now */
+    Success = Hive->FileWrite(Hive,
+                              HFILE_TYPE_PRIMARY,
+                              &FileOffset,
+                              LogHeader,
+                              Hive->Cluster * HSECTOR_SIZE);
+    if (!Success)
+    {
+        DPRINT1("HvpRecoverHeaderFromLog(): Couldn't write the base header to primary hive!\n");
+        Hive->Free(LogHeader, Hive->BaseBlockAlloc);
+        return Fail;
+    }
+
+    *BaseBlock = LogHeader;
+    return HeaderResuscitated ? SelfHeal : HiveSuccess;
+}
+
+/**
+ * @brief
+ * Recovers the registry data by obtaining it
+ * from a log that is associated with the hive.
+ *
+ * @param[in] Hive
+ * A pointer to a hive descriptor associated
+ * with the log file where the hive data is to
+ * be read from.
+ *
+ * @param[in] BaseBlock
+ * A pointer to a base block header.
+ *
+ * @return
+ * Returns HiveSuccess if the data was obtained
+ * normally from the log. Fail is returned if
+ * self-healing is disabled and we couldn't be
+ * able to read the data from the log or the
+ * dirty vector signature is garbage or we
+ * failed to write the data block to the primary
+ * hive. SelfHeal is returned to indicate that
+ * the log is corrupt and the system will continue
+ * to be recovered at the expense of data loss.
+ */
+RESULT
+CMAPI
+HvpRecoverDataFromLog(
+    _In_ PHHIVE Hive,
+    _In_ PHBASE_BLOCK BaseBlock)
+{
+    BOOLEAN Success;
+    ULONG FileOffset;
+    ULONG BlockIndex;
+    ULONG LogIndex;
+    ULONG StorageLength;
+    UCHAR DirtyVector[HSECTOR_SIZE];
+    UCHAR Buffer[HBLOCK_SIZE];
+
+    /* Read the dirty data from the log */
+    FileOffset = HV_LOG_HEADER_SIZE;
+    Success = Hive->FileRead(Hive,
+                             HFILE_TYPE_LOG,
+                             &FileOffset,
+                             DirtyVector,
+                             HSECTOR_SIZE);
+    if (!Success)
+    {
+        /*
+         * We failed? See if self-heal mode is
+         * enabled on this system.
+         */
```

Comment:
again a 4 line-comment-block that does not add **any** information at all. Strip it, help your readers and their mousewheels.